app/team.go

```diff
@@ -447,10 +447,32 @@ func (a *App) AddUserToTeamByToken(userId string, tokenId string) (*model.Team,
 	}
 	user := result.Data.(*model.User)
 
+	if user.IsGuest() && token.Type == TOKEN_TYPE_TEAM_INVITATION {
+		return nil, model.NewAppError("AddUserToTeamByToken", "api.user.create_user.invalid_invitation_type.app_error", nil, "", http.StatusBadRequest)
+	}
+	if !user.IsGuest() && token.Type == TOKEN_TYPE_GUEST_INVITATION {
+		return nil, model.NewAppError("AddUserToTeamByToken", "api.user.create_user.invalid_invitation_type.app_error", nil, "", http.StatusBadRequest)
+	}
```

Comment:
No, the problem is about the expectations of privacy. If you invite somebody as a guest, you expect that guy to be a guest, and have restricted access to your team. So, as soon as is a regular user he shouldn't be able to accept a guest invite because it can generate a situation where the user see more than expected.